[tools] Rewrite Ruby scripts in Python for consistency

meenbeese opened the pull request:
- Rewrote the scripts in Python with same functionality
- Relocated the scripts to be under `tools/python`
- Tested them with the latest Python 3.13 version

meenbeese commented:
@rtsisyk

vng commented:
Should run "Validate translation strings" when tools/unix; tools/python scripts were changed.

meenbeese commented:
> Should run "Validate translation strings" when tools/unix; tools/python scripts were changed.

Do you mean the GitHub actions workflow? It only runs when tools/python/string_utils.py is changed, but I can make a PR to fix that if you want.

vng commented:
Ah, my bad. I thought that check_consistency.py is called in some Actions, but not.
Looks like this script does a good check, could you please copy its output here after your refactoring?

pastk commented on docs/TRANSLATIONS.md:
```diff
@@ -60,7 +60,7 @@ Check `tools/python/strings_utils.py -h` to see all of them.
 
 To check consistency of types_strings.txt with categories.txt run:
```
But the script operates with `string.txt`...

Could you please add a comment to the script describing what exactly does it check?
I've tried to make some changes to the UI search categories in categories.txt but couldn't make the script fail.

meenbeese replied:
Added a comment in the script and updated the documentation.

meenbeese commented on tools/python/category_consistency/check_consistency.py:
It is already executable as it has a shebang.

pastk replied:
Its not enough, the execution permission must be set e.g. `chmod a+x` so that it could use shebang and be executed via `./tools/python/category_consistency/check_consistency.py`

meenbeese replied:
The execution permission is set by the user, and not by the script itself.

pastk replied:
The repo holds permissions for each file. Most of our scripts are set to `-rwxrwxr-x` (but not all - its inconsistent atm). Just execute `chmod a+x tools/python/category_consistency/check_consistency.py` and `git status` will report the file as modified.

meenbeese replied:
Ok, I ran the chmod command, but it is not like I can commit this to the repo. So I don't really understand what you are trying to do here.

pastk commented:
Still the script outputs "Success: All categories are consistent." always.

E.g. I go to `@category_transport` and change `en:5Transport` to something else.
Or I go to `[category_fuel]` and change `en = Gas` to something else.
It report "Success" anyway.

meenbeese commented:
I fixed the parsing logic and implemented a way to properly display the results. This is what I am getting now:
```
$ python .\tools\python\category_consistency\check_consistency.py

Missing categories:
  - back
  - cancel
  - cancel_download
  - delete
  - download_maps
  - download_has_failed
  - downloading
  - kilometres
  - leave_a_review
  - mb
  - gb
  - miles
  - core_my_position
  - later
  - search
  - search_map
  - use_cellular_data
  - location_is_disabled_long_text
  - zoom_to_country
  - country_status_download_without_size
  - country_status_download_failed
  - try_again
  - about_menu_title
  - about_headline
  - about_proposition_1
  - about_proposition_2
  - about_proposition_3
  - about_developed_by_enthusiasts
  - location_settings
  - close
  - unsupported_phone
  - download
  - disconnect_usb_cable
  - not_enough_free_space_on_sdcard
  - download_resources
  - download_resources_continue
  - downloading_country_can_proceed
  - download_country_ask
  - update_country_ask
  - pause
  - continue_button
  - download_country_failed
  - add_new_set
  - bookmark_set_name
  - bookmark_sets
  - bookmarks
  - bookmarks_and_tracks
  - core_my_places
  - name
  - address
  - list
  - settings
  - maps_storage
  - maps_storage_summary
  - maps_storage_downloaded
  - maps_storage_internal
  - maps_storage_shared
  - maps_storage_removable
  - maps_storage_external
  - maps_storage_free_size
  - move_maps
  - move_maps_error
  - wait_several_minutes
  - measurement_units
  - measurement_units_summary
  - category_eat
  - category_food
  - category_transport
  - category_fuel
  - category_parking
  - category_shopping
  - category_secondhand
  - category_hotel
  - category_tourism
  - category_entertainment
  - category_atm
  - category_nightlife
  - category_children
  - category_bank
  - category_pharmacy
  - category_hospital
  - category_toilet
  - category_post
  - category_police
  - category_wifi
  - category_recycling
  - category_water
  - category_rv
  - description
  - share_bookmarks_email_subject
  - share_bookmarks_email_body
  - load_kmz_title
  - load_kmz_successful
  - load_kmz_failed
  - unknown_file_type
  - failed_to_open_file
  - edit
  - unknown_current_position
  - cant_change_this_setting
  - downloading_is_active
  - my_position_share_sms
  - bookmark_share_email_subject
  - my_position_share_email_subject
  - my_position_share_email
  - share
  - email
  - copied_to_clipboard
  - info
  - done
  - version
  - data_version
  - osm_log_out_confirmation
  - tracks_title
  - length
  - share_my_location
  - prefs_group_general
  - prefs_group_information
  - prefs_group_route
  - pref_zoom_title
  - pref_zoom_summary
  - pref_map_style_title
  - pref_appearance_title
  - pref_appearance_light
  - pref_appearance_dark
  - off
  - on
  - auto
  - pref_map_3d_title
  - pref_map_3d_buildings_title
  - pref_map_3d_buildings_disabled_summary
  - pref_tts_enable_title
  - pref_tts_street_names_title
  - pref_tts_street_names_description
  - pref_tts_language_title
  - pref_tts_test_voice_title
  - pref_tts_playing_test_voice
  - pref_tts_unavailable
  - pref_tts_other_section_title
  - pref_map_auto_zoom
  - placepage_distance
  - search_show_on_map
  - menu
  - website
  - news
  - github
  - telegram
  - matrix
  - mastodon
  - facebook
  - twitter
  - instagram
  - vk
  - editor_line_social_network
  - openstreetmap
  - feedback
  - rate_the_app
  - help
  - faq
  - donate
  - donate_description
  - already_donated
  - remind_me_later
  - support_organic_maps
  - how_to_support_us
  - copyright
  - report_a_bug
  - report_incorrect_map_bug
  - volunteer
  - follow_us
  - email_error_body
  - email_error_title
  - pref_calibration_title
  - compass_calibration_recommended
  - compass_calibration_required
  - long_tap_toast
  - downloader_update_all_button
  - downloader_cancel_all
  - downloader_downloaded_subtitle
  - downloader_available_maps
  - downloader_queued
  - downloader_near_me_subtitle
  - downloader_status_maps
  - downloader_download_all_button
  - downloader_downloading
  - downloader_search_results
  - downloader_status_outdated
  - downloader_status_failed
  - downloader_delete_map_while_routing_dialog
  - routing_failed_cross_mwm_building
  - downloader_download_map
  - downloader_retry
  - downloader_delete_map
  - downloader_update_map
  - google_play_services
  - pref_use_google_play
  - routing_download_maps_along
  - routing_requires_all_map
  - routing_not_enough_space
  - bookmark
  - enable_location_services
  - save
  - edit_description_hint
  - create
  - red
  - yellow
  - blue
  - green
  - purple
  - orange
  - brown
  - pink
  - deep_purple
  - light_blue
  - cyan
  - teal
  - lime
  - deep_orange
  - gray
  - blue_gray
  - dialog_routing_disclaimer_title
  - dialog_routing_disclaimer_priority
  - dialog_routing_disclaimer_precision
  - dialog_routing_disclaimer_recommendations
  - dialog_routing_disclaimer_borders
  - dialog_routing_disclaimer_beware
  - dialog_routing_check_gps
  - dialog_routing_error_location_not_found
  - dialog_routing_location_turn_wifi
  - dialog_routing_location_turn_on
  - dialog_routing_location_unknown_turn_on
  - dialog_routing_download_files
  - dialog_routing_download_and_update_all
  - dialog_routing_unable_locate_route
  - dialog_routing_cant_build_route
  - dialog_routing_change_start_or_end
  - dialog_routing_change_start
  - dialog_routing_start_not_determined
  - dialog_routing_select_closer_start
  - dialog_routing_change_end
  - dialog_routing_end_not_determined
  - dialog_routing_select_closer_end
  - dialog_routing_change_intermediate
  - dialog_routing_intermediate_not_determined
  - dialog_routing_system_error
  - dialog_routing_application_error
  - dialog_routing_try_again
  - not_now
  - dialog_routing_download_and_build_cross_route
  - dialog_routing_download_cross_route
  - search_without_internet_advertisement
  - search_select_map
  - show
  - hide
  - routing_planning_error
  - routing_arrive
  - dialog_routing_download_and_update_maps
  - categories
  - history
  - search_not_found
  - search_not_found_query
  - search_history_title
  - search_history_text
  - clear_search
  - read_in_wikipedia
  - wikimedia_commons
  - p2p_your_location
  - p2p_start
  - p2p_from_here
  - p2p_to_here
  - p2p_only_from_current
  - p2p_reroute_from_current
  - next_button
  - editor_time_from
  - editor_time_to
  - editor_time_add
  - editor_time_delete
  - editor_time_allday
  - editor_time_open
  - editor_time_close
  - editor_time_add_closed
  - editor_time_title
  - editor_time_advanced
  - editor_time_simple
  - editor_hours_closed
  - editor_example_values
  - editor_correct_mistake
  - editor_add_select_location
  - editor_done_dialog_1
  - share_with_friends
  - editor_report_problem_desription_1
  - editor_report_problem_desription_2
  - editor_report_problem_send_button
  - editor_report_problem_title
  - editor_report_problem_no_place_title
  - editor_report_problem_under_construction_title
  - editor_report_problem_duplicate_place_title
  - autodownload
  - closed_now
  - daily
  - twentyfour_seven
  - day_off_today
  - day_off
  - today
  - opens_tomorrow_at
  - opens_dayoftheweek_at
  - opens_at
  - opens_in
  - closes_at
  - closes_in
  - closed
  - add_opening_hours
  - edit_opening_hours
  - no_osm_account
  - register_at_openstreetmap
  - password_8_chars_min
  - invalid_username_or_password
  - login
  - not_signed_in
  - login_osm
  - password
  - forgot_password
  - osm_account
  - logout
  - last_upload
  - thank_you
  - edit_place
  - place_name
  - add_language
  - street
  - house_number
  - details
  - add_street
  - empty_street_name_error
  - choose_language
  - choose_street
  - postal_code
  - cuisine
  - select_cuisine
  - email_or_username
  - phone
  - editor_add_phone
  - level
  - level_value_generic
  - please_note
  - downloader_delete_map_dialog
  - downloader_update_maps
  - downloader_mwm_migration_dialog
  - downloader_search_field_hint
  - migration_download_error_dialog
  - common_check_internet_connection_dialog
  - downloader_no_space_title
  - downloader_no_space_message
  - editor_login_error_dialog
  - editor_profile_changes
  - editor_focus_map_on_location
  - editor_add_select_category
  - editor_add_select_category_popular_subtitle
  - editor_add_select_category_all_subtitle
  - editor_edit_place_title
  - editor_add_place_title
  - editor_edit_place_name_hint
  - editor_default_language_hint
  - editor_edit_place_category_title
  - detailed_problem_description
  - editor_report_problem_other_title
  - placepage_add_business_button
  - whatsnew_editor_message_1
  - dialog_incorrect_feature_position
  - message_invalid_feature_position
  - osm_presentation
  - login_osm_presentation
  - login_to_make_edits_visible
  - migration_no_space_message
  - editor_sharing_title
  - downloader_of
  - download_over_mobile_header
  - download_over_mobile_message
  - error_enter_correct_house_number
  - editor_storey_number
  - error_enter_correct_storey_number
  - editor_zip_code
  - error_enter_correct_zip_code
  - core_placepage_unknown_place
  - editor_other_info
  - editor_note_hint
  - editor_about_osm
  - editor_detailed_description_hint
  - editor_detailed_description
  - editor_more_about_osm
  - editor_operator
  - operator
  - editor_category_unsuitable_title
  - editor_category_unsuitable_text
  - downloader_no_downloaded_maps_title
  - downloader_no_downloaded_maps_message
  - current_location_unknown_error_title
  - current_location_unknown_error_message
  - location_services_disabled_header
  - location_services_disabled_message
  - location_services_disabled_1
  - location_services_disabled_2
  - location_services_disabled_3
  - location_services_disabled_on_device_2
  - location_services_disabled_on_device_3
  - location_services_disabled_on_device_4
  - location_services_disabled_on_device_additional_message
  - m
  - km
  - kilometers_per_hour
  - mi
  - ft
  - miles_per_hour
  - hour
  - minute
  - day
  - placepage_more_button
  - book_button
  - more_on_kayak
  - dialog_kayak_disclaimer
  - dialog_kayak_button
  - placepage_call_button
  - placepage_edit_bookmark_button
  - placepage_bookmark_name_hint
  - placepage_personal_notes_hint
  - placepage_delete_bookmark_button
  - editor_edits_sent_message
  - editor_comment_hint
  - editor_reset_edits_message
  - editor_reset_edits_button
  - editor_remove_place_message
  - editor_remove_place_button
  - editor_place_doesnt_exist
  - delete_place_empty_comment_error
  - text_more_button
  - error_enter_correct_phone
  - error_enter_correct_web
  - error_enter_correct_email
  - error_enter_correct_facebook_page
  - error_enter_correct_instagram_page
  - error_enter_correct_twitter_page
  - error_enter_correct_vk_page
  - error_enter_correct_line_page
  - refresh
  - placepage_add_place_button
  - editor_share_to_all_dialog_title
  - editor_share_to_all_dialog_message_1
  - editor_share_to_all_dialog_message_2
  - navigation_stop_button
  - recent_track_background_dialog_title
  - off_recent_track_background_button
  - sharing_call_action_look
  - recent_track_background_dialog_message
  - track_recording
  - general_settings
  - accept
  - decline
  - search_in_table
  - mobile_data_dialog
  - mobile_data_option_always
  - mobile_data_option_today
  - mobile_data_option_not_today
  - mobile_data
  - mobile_data_description
  - mobile_data_option_never
  - mobile_data_option_ask
  - traffic_update_maps_text
  - big_font
  - traffic_update_app
  - traffic_update_app_message
  - traffic_data_unavailable
  - enable_logging
  - log_file_size
  - feedback_general
  - prefs_languages_information
  - prefs_languages_information_off
  - prefs_languages_information_off_link
  - transliteration_title
  - learn_more
  - core_exit
  - routing_add_start_point
  - routing_add_finish_point
  - planning_route_manage_route
  - button_plan
  - placepage_remove_stop
  - planning_route_remove_title
  - placepage_add_stop
  - start_from_my_position
  - alert_reauth_message
  - alert_reauth_message_ios
  - alert_reauth_link_text_ios
  - dialog_error_storage_title
  - dialog_error_storage_message
  - setting_emulate_bad_storage
  - core_entrance
  - error_enter_correct_name
  - bookmark_lists
  - bookmark_lists_hide_all
  - bookmark_lists_show_all
  - bookmarks_places
  - bookmarks_create_new_group
  - bookmarks_import
  - downloader_hide_screen
  - downloader_percent
  - downloader_process
  - downloader_applying
  - bookmarks_error_message_share_general
  - bookmarks_error_title_share_empty
  - bookmarks_error_message_share_empty
  - bookmarks_error_title_empty_list_name
  - bookmarks_error_message_empty_list_name
  - bookmarks_new_list_hint
  - bookmarks_error_title_list_name_already_taken
  - bookmarks_error_message_list_name_already_taken
  - bookmarks_error_title_list_name_too_long
  - please_wait
  - phone_number
  - profile
  - bookmarks_detect_title
  - bookmarks_detect_message
  - button_convert
  - bookmarks_convert_error_title
  - bookmarks_convert_error_message
  - restore
  - tracks
  - privacy
  - privacy_policy
  - terms_of_use
  - button_layer_traffic
  - button_layer_subway
  - layers_title
  - subway_data_unavailable
  - bookmarks_empty_list_title
  - bookmarks_empty_list_message
  - category_desc_more
  - title_error_downloading_bookmarks
  - popular_place
  - export_file
  - export_file_gpx
  - delete_list
  - hide_from_map
  - public_access
  - limited_access
  - bookmark_list_description_hint
  - not_shared
  - tags_loading_error_subtitle
  - download_button
  - speedcams_alert_title
  - place_description_title
  - notification_channel_downloader
  - pref_tts_speedcams_auto
  - pref_tts_speedcams_always
  - pref_tts_speedcams_never
  - power_managment_title
  - power_managment_description
  - power_managment_setting_never
  - power_managment_setting_auto
  - power_managment_setting_manual_max
  - enable_logging_warning_message
  - access_rules_author_only
  - driving_options_title
  - avoid_tolls
  - avoid_unpaved
  - avoid_ferry
  - avoid_motorways
  - unable_to_calc_alert_title
  - unable_to_calc_alert_subtitle
  - define_to_avoid_btn
  - change_driving_options_btn
  - toll_road
  - unpaved_road
  - ferry_crossing
  - trip_start
  - pick_destination
  - follow_my_position
  - search_results
  - then_turn
  - redirect_route_alert
  - yes
  - no
  - yes_available
  - no_available
  - capacity
  - network
  - trip_finished
  - keyboard_availability_alert
  - dialog_routing_change_start_carplay
  - dialog_routing_change_end_carplay
  - dialog_routing_check_gps_carplay
  - dialog_routing_unable_locate_route_carplay
  - dialog_routing_download_files_carplay
  - dialog_routing_system_error_carplay
  - dialog_routing_rebuild_from_current_location_carplay
  - dialog_routing_rebuild_for_vehicle_carplay
  - not_all_shown_bookmarks_carplay
  - switch_to_phone_bookmarks_carplay
  - ok
  - speedcams_alert_title_carplay_1
  - speedcams_alert_title_carplay_2
  - download_map_carplay
  - carplay_roundabout_exit
  - sort
  - sort_bookmarks
  - sort_default
  - sort_type
  - sort_distance
  - sort_date
  - sort_name
  - by_default
  - by_type
  - by_distance
  - by_date
  - by_name
  - week_ago_sorttype
  - month_ago_sorttype
  - moremonth_ago_sorttype
  - moreyear_ago_sorttype
  - near_me_sorttype
  - others_sorttype
  - food_places
  - tourist_places
  - museums
  - parks
  - swim_places
  - mountains
  - animals
  - hotels
  - buildings
  - money
  - shops
  - parkings
  - fuel_places
  - medicine
  - search_in_the_list
  - religious_places
  - select_list
  - transit_not_found
  - dialog_pedestrian_route_is_long_header
  - dialog_pedestrian_route_is_long_message
  - button_layer_isolines
  - isolines_activation_error_dialog
  - isolines_location_error_dialog
  - elevation_profile_diff_level
  - elevation_profile_diff_level_easy
  - elevation_profile_diff_level_moderate
  - elevation_profile_diff_level_hard
  - elevation_profile_ascent
  - elevation_profile_descent
  - elevation_profile_minaltitude
  - elevation_profile_maxaltitude
  - elevation_profile_difficulty
  - elevation_profile_distance
  - elevation_profile_time
  - isolines_toast_zooms_1_10
  - downloader_updating_ios
  - downloader_loading_ios
  - download_map_title
  - disk_error
  - disk_error_title
  - connection_failure
  - disconnect_usb_cable_title
  - enable_keep_screen_on
  - enable_keep_screen_on_description
  - enable_show_on_lock_screen
  - enable_show_on_lock_screen_description
  - change_map_locale
  - whats_new_auto_update_title
  - whats_new_auto_update_message
  - whats_new_auto_update_button_size
  - whats_new_auto_update_button_later
  - placepage_delete_track_button
  - placepage_track_name_hint
  - move
  - track_title
  - splash_subtitle
  - telegram_url
  - instagram_url
  - tts_info_link
  - translated_om_site_url
  - osm_wiki_about_url
  - comma_separated_pair
  - app_tip_00
  - app_tip_01
  - app_tip_02
  - app_tip_03
  - app_tip_04
  - app_tip_05
  - app_tip_06
  - app_tip_07
  - app_tip_08
  - app_tip_09
  - car_used_on_the_phone_screen
  - car_used_on_the_car_screen
  - aa_connected_title
  - car_continue_on_the_phone
  - car_continue_in_the_car
  - aa_location_permissions_request
  - aa_grant_permissions
  - aa_connected_to_car_notification_title
  - button_layer_outdoor
  - browser_not_available
  - volume
  - bookmarks_export
  - change_color
  - pref_tts_open_system_settings
  - pref_tts_no_system_tts
  - map
  - drive_through
  - clear_the_search
  - zoom_in
  - zoom_out
  - website_menu
  - view_menu
  - bugreport_alert_message
  - enable_icloud_synchronization_title
  - enable_icloud_synchronization_message
  - icloud_disabled_title
  - icloud_disabled_message
  - enable
  - backup
  - icloud_synchronization_error_alert_title
  - icloud_synchronization_error_connection_error
  - icloud_synchronization_error_quota_exceeded
  - icloud_synchronization_error_cloud_is_unavailable
  - open_in_app
  - self_service
  - select_option
  - outdoor_seating
  - power_save_dialog_summary
  - bookmarks_recently_deleted
  - clear
  - delete_all
  - recover
  - recover_all
  - start_track_recording
  - stop_track_recording
  - stop_without_saving
  - continue_recording
  - track_recording_alert_title
  - track_recording_toast_nothing_to_save
  - error_no_file_manager_app
  - choose_color
  - edit_track
  - uri_open_location_failed
  - nav_auto
Failure: Inconsistencies found in category definitions.
```

meenbeese commented:
I also moved the files to under the `category` namespace so that related scripts can be added there in the future.

strump commented on tools/python/category/omim_parsers.py:
```diff
@@ -0,0 +1,98 @@
+                    continue
+
+                # Match a new category
+                category_match = self.category().match(line)
```
Small optimization here: no need to call `self.category()` for every line.

Please create new variable at the start of `parse_file(...)` method and put result of `self.category()`
For example:

```python
category_regex = self.category()
with open(...):
    ...
    category_match = category_regex.match(line)
```

strump commented on tools/python/category/check_consistency.py:
```diff
@@ -0,0 +1,123 @@
+    return [cat.strip().strip('"') for cat in raw_categories.split(",")]
+
+
+def compare_categories(string_cats, search_cats, cpp_cats):
```
Please add type annotations like you did in omim_parsers.py

strump commented on tools/python/category/check_consistency.py:
```diff
@@ -0,0 +1,123 @@
+    missing_categories = []
+    extra_categories = []
+
+    for category_name in search_cats.keys():
+        if category_name not in string_cats:
+            missing_categories.append(category_name)
```
You can simplify this code by using `set` collection. For example:

```python
missing_categories = search_cats.keys() - set(string_cats)
```

strump commented on tools/python/category/check_consistency.py:
```diff
@@ -0,0 +1,123 @@
+        if category_name not in string_cats:
+            missing_categories.append(category_name)
+
+    for cpp_cat in cpp_cats:
+        if cpp_cat not in search_cats:
+            extra_categories.append(cpp_cat)
```
You can simplify this code by using `set` collection. For example:

```python
extra_categories = set(cpp_cats) - set(search_cats)
```

pastk commented:
> I fixed the parsing logic and implemented a way to properly display the results. This is what I am getting now:

The output is erroneous still...

> 
> Missing translations for categories in strings.txt:
>   - amenity-atm
>   - amenity-bank

They're not supposed to be present in `strings.txt` at all. Only the ones from `displayed_categories.cpp`.

> 
> Extra categories found in displayed_categories.cpp but not in categories.txt:
>   - category_eat

But all of them are there, e.g. 
```
# First keyword should match [category_eat] definition in strings.txt!
@category_eat
en:Where to eat|eat|Food
```

pastk reviewed (changes requested):
As commented above need to fix the check logic..
